### PR TITLE
4789 item variant UI pt2

### LINKED ISSUE #4789: GAPS: Implement item variants: Item UI changes
Part of epic #4362

## Is your feature request related to a problem? Please describe 👀
There are several aspects of the GAPS work which require item variant capabilities
For reference this is the full [item variant PRD](https://docs.google.com/document/d/1fRZslRohRGaZnUo13a2eKZjfO0DGWwXtIFsbEah7A4Y/edit#heading=h.hyld495fuxe5)

We won't need all aspects of that; but we will need the ability to create and edit and sync item variants.

The variants will need to have these properties for the GAPS work:

| Name | Type   | Reference |
|:------ |:------ |:---- |
| Volume (x3) | float | #3917 |
| Bundle items | reference to n item variants | #3923 |
| Temperature range | enum | #4368 |

## Describe the solution you'd like 🎁

See #5141

UI - on OMS central:

image

### Handling stock

Done in separate issue: #5145

### PR DESCRIPTION
**Pull request by lache-melvin**
Fixes #4789

# 👩🏻‍💻 What does this PR do?

Moves add/edit of item variants to modal:
Screenshot 2024-10-25 at 10 30 27 AM

## 💌 Any notes for the reviewer?

# 🧪 Testing

- [ ] No backend wired up, but check add/edit open the modal as expected, and send the right payload to the backend

# 📃 Documentation

- [x] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

File: client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemPackagingVariantsTable.tsx [hunk 1]
```diff
@@ -6,21 +6,54 @@ import {
...
-    { key: 'packSize', Cell: TooltipTextCell, label: 'label.pack-size' },
     {
-      key: 'volumePerUnit',
+      key: 'packagingLevel',
       Cell: TooltipTextCell,
+      label: 'label.level',
+    },
+    {
+      key: 'name',
+      Cell: update ? TextInputCell : TooltipTextCell,
```
Comment (lache-melvin) [inline client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemPackagingVariantsTable.tsx:42]: thoughts on logic here to be able to reuse the table component?

File: client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemVariantsTab.tsx [hunk 2]
```diff
@@ -16,134 +14,81 @@ import {
...
-            Input={
-              <BasicTextInput
-                value={variant.name}
-                onChange={event => {
-                  setVariant({ ...variant, name: event.target.value });
-                }}
-                fullWidth
-              />
-            }
+            Input={<BasicTextInput value={variant.name} disabled fullWidth />}
```
Comment (lache-melvin) [inline client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemVariantsTab.tsx:91]: Could have reused the edit form from the modal here, but am under the impression that item variants are likely to end up with a lot of fields, and probably only want a subset in the tab view 👀

File: client/packages/system/src/Item/api/hooks/useItemVariant/useItemVariant.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        },
+      ],
+    }
+  );
+
+  const draftRef = useRef(draft);
+
+  useEffect(() => {
+    draftRef.current = draft;
+  }, [draft]);
```
Comment (lache-melvin) [inline client/packages/system/src/Item/api/hooks/useItemVariant/useItemVariant.ts:63]: Not stoked with this, I don't think it should be required - there's something up with the packaging table, I think the `setter` is a closure over `updateDraft` that had stale draft state. Was a rabbit hole so just using a ref here to ensure packaging variant table updates don't clear the rest of the draft state, but would like to understand it more...

File: client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemPackagingVariantsTable.tsx (hunk 1 above)
Comment (CarlosNZ) [reply client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemPackagingVariantsTable.tsx:37]: Seems good 👍🏼

File: client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemVariantsTab.tsx (hunk 2 above)
Comment (CarlosNZ) [reply client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemVariantsTab.tsx:91]: Really? I would have thought we'd want to see the same fields on the tab view as in the modal. How many are we talking?

**Review by CarlosNZ — APPROVED** [review]
Looks good, allowing for the TO-DOs and the other considerations we discussed IRL.

Maybe just chuck in the empty array when no variants defined, and add the "No data" component as discussed?

File: client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemVariantsTab.tsx (hunk 2 above)
Comment (lache-melvin) [reply client/packages/system/src/Item/DetailView/Tabs/ItemVariants/ItemVariantsTab.tsx:91]: I think we're talking long term, but that's what Adam has said 😁 I think I'll stick with it anyway - nice not to do things like query all manufacturers for the dropdown, only to display read only data

**Pull request merged**